src/sctools/fastq.py

```diff
@@ -0,0 +1,197 @@
+        else:
+            self._data[1] = value
+
+    @property
+    def name2(self):
+        return self._data[2]
+
+    @name2.setter
+    def name2(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('name2 must be str or bytes')
+        else:
+            self._data[2] = value
+
+    @property
+    def quality(self):
+        return self._data[3]
+
+    @quality.setter
+    def quality(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('quality must be str or bytes')
+        else:
+            self._data[3] = value
+
+    def __bytes__(self):
+        try:
+            return b''.join(self._data)
+        except TypeError:
+            return ''.join(self._data).encode()
+
+    def __str__(self):
+        try:
+            return ''.join(self._data)
+        except TypeError:
+            return b''.join(self._data).decode()
+
+    def __repr__(self):
+        return "Name: %s\nSequence: %s\nName2: %s\nQuality: %s\n" % (
+            self._data[0], self._data[1], self._data[2], self._data[3])
+
+    def __len__(self):
+        return len(self.sequence)
+
+    def average_quality(self):
+        raise NotImplementedError
+
+
+class BytesRecord(Record):
+    """Fastq record object
+
+    Defines several properties for accessing fastq record information:
+    :property bytes name: name field
+    :property bytes sequence: sequence field
+    :property bytes name2: second name field
+    :property bytes quality: quality field
+
+    Also defines several methods for accessing SEQC annotation fields:
+    :property float average_quality: return the mean quality of FastqRecord
+    """
+
+    def average_quality(self):
+        """return the average quality of this record"""
+        return sum(c for c in self.quality[:-1]) / (len(self.quality[:-1]) - 1) - 33
+
+
+class StrRecord(Record):
+    """Fastq record object
+
+    Defines several properties for accessing fastq record information:
+    :property str name: name field
+    :property str sequence: sequence field
+    :property str name2: second name field
+    :property str quality: quality field
+
+    Utilities for parsing read attributes:
+    :property average_quality: return the mean quality of FastqRecord
+    """
+
+    def average_quality(self):
+        """calculate the average quality of the fastq read
+
+        :return float: average quality of fastq read
+        """
+        b = self.quality[:-1].encode()
+        return sum(c for c in b) / len(b) - 33
+
+
+class Reader(reader.Reader):
+    """
+    Fastq Reader, defines some special methods for reading and summarizing fastq data:
+
+    :method __iter__: Iterator over fastq Record objects
+    :method __len__: return number of records in file
+    :method estimate_sequence_length: estimate the length of fastq sequences in file
+    """
+
+    @staticmethod
+    def record_grouper(iterable):
```

Comment:
I understand this function better after talking in person. A comment about its purpose would probably be helpful, here or maybe in the class docstring.